versions-common/src/main/java/org/codehaus/mojo/versions/api/VersionRetrievalException.java

```diff
@@ -69,12 +81,30 @@ public VersionRetrievalException(Throwable cause) {
      *
      * @param  message the detail message (which is saved for later retrieval
      *         by the {@link #getMessage()} method).
+     * @param artifact {@link Artifact} instance causing the problem
      * @param  cause the cause (which is saved for later retrieval by the
      *         {@link #getCause()} method).  (A {@code null} value is
      *         permitted, and indicates that the cause is nonexistent or
      *         unknown.)
      */
-    public VersionRetrievalException(String message, Throwable cause) {
+    public VersionRetrievalException(String message, Artifact artifact, Throwable cause) {
         super(message, cause);
+        setArtifact(artifact);
+    }
+
+    /**
+     * Returns the artifact causing the problem with version retrieval, if available
+     * @return {@link Optional} object containing artifact causing the problem with version retrieval, or empty
+     */
+    public Optional<Artifact> getArtifact() {
+        return Optional.ofNullable(artifact);
+    }
+
+    /**
+     * Sets the artifact causing the problem with version retrieval
+     * @param artifact {@link Artifact} instance causing the problem
+     */
+    public void setArtifact(Artifact artifact) {
```

Comment:
exception should be immutable, so I would like to declare artifact as final and use only constructors ...

Reply:
Done